### PR TITLE
Integrate completion into ArgParse

### PR DESCRIPTION
**Pull request by t-wissmann**
This introduces the new type CallOrComplete that combines the invocation
and argument completion of a command.

Effectively,  this will make commands functions with reasonable
parameter types and names and the CallOrComplete-function acts as
glueing code or binding to the user-accessible command interface.

To illustrate the new functionality, it is applied to the commands
bring, focus, shift, tag_status, remove_monitor whose test suite
regarding parse errors and completion is respectively extended.

The word-list suggestion parameter is not yet used, but it will be
useful for commands like cycle_all (completion for -1 and +1) or resize
(completion for +0.02 and -0.02), which will be handled in separate
commits.

### CODE REVIEW

File: src/command.h [hunk 1]
```diff
@@ -91,6 +91,36 @@ class CommandBinding {
...
+
+    template <typename ClassName>
+    CommandBinding(ClassName* object,
+                   void(ClassName::*member)(CallOrComplete))
+        : CommandBinding(std::bind(member, object,
+                            std::placeholders::_1))
+    {
+    }
+
     // FIXME: Remove after C++ transition
```
Comment (blarz) [inline src/command.h:124]: Not related to this PR, but Is this possible yet?

**Review by blarz — APPROVED** [review]
This is pretty cool (not sure if I understood all of it, but tests still pass, so..)

File: src/command.h (hunk 1 above)
Comment (t-wissmann) [reply src/command.h:124]: There are still a few in main.cpp, but I hope I can migrate them soon.

**Comment by t-wissmann** [comment]
> This is pretty cool (not sure if I understood all of it, but tests still pass, so..)

Thanks :) (The complexity is a real downside, so I'm not sure whether I would use it for all commands, because for complex commands like `rule` it maybe makes things more complicated than necessary)

**Pull request merged**